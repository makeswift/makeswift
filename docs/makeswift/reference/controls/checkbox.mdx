---
title: Checkbox
description: Adds a Checkbox panel in the Makeswift builder to visually edit a `boolean` prop.
icon: "square-check"
---

<Frame
  type="glass"
  caption="A checkbox panel on a Navigation component to show the logo"
>
  <img src="/images/checkbox-panel.png" />
</Frame>

## Params

<ParamField query="label" type="string" default="Label">
  Text for the panel label in the Makeswift builder.
</ParamField>

<ParamField query="defaultValue" type="boolean">
  The value passed to your component when nothing is set in the Makeswift
  builder.
</ParamField>

## Prop type

The Checkbox control passes a `boolean` to your component. If you don't set a `defaultValue` and no value is set in the builder, your component receives `undefined`.

## Example

The following example adds a Checkbox control to the `showLogo` prop of a Navigation component.

<CodeGroup>

```tsx Navigation.makeswift.ts
import { Checkbox } from "@makeswift/runtime/controls"

import { runtime } from "@/makeswift/runtime"

import { Navigation } from "./Navigation"

runtime.registerComponent(Navigation, {
  type: "navigation",
  label: "Navigation",
  props: {
    showLogo: Checkbox({
      label: "Show Logo",
      defaultValue: false,
    }),
  },
})
```

```tsx Navigation.tsx
interface Props {
  showLogo: boolean
}

export function Navigation({ showLogo }: Props) {
  return (
    <nav>
      {props.showLogo && <img src="/logo.png" />}
      {/* ... */}
    </nav>
  )
}
```

</CodeGroup>

import NamingConventions from "/snippets/naming-conventions.mdx"

<NamingConventions />
