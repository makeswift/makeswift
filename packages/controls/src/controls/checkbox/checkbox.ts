import { match, P } from 'ts-pattern'
import { z } from 'zod'

import { safeParse, type ParseResult } from '../../lib/zod'

import { ControlDataTypeKey } from '../../common'
import { type CopyContext } from '../../context'
import { type DeserializedRecord } from '../../serialization'

import {
  ControlDefinition,
  type Resolvable,
  type SchemaType,
} from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'
import { ControlDefinitionVisitor } from '../visitor'

type Config = z.infer<typeof Definition.schema.relaxed.config>

type SchemaByDefaultValue<D extends Config['defaultValue']> =
  undefined extends D
    ? typeof Definition.schema.relaxed
    : typeof Definition.schema.strict

type Schema<C extends Config> = SchemaByDefaultValue<C['defaultValue']>
type DataType<C extends Config> = z.infer<Schema<C>['data']>
type ValueType<C extends Config> = z.infer<Schema<C>['value']>
type ResolvedValueType<C extends Config> = z.infer<Schema<C>['resolvedValue']>

type ReturnedSchemaType<C extends Config> = {
  definition: typeof Definition.schema.relaxed.definition
  type: typeof Definition.schema.relaxed.type
  data: SchemaType<DataType<C>>
  value: SchemaType<ValueType<C>>
  resolvedValue: SchemaType<ResolvedValueType<C>>
}

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>
> {
  private static readonly v1DataType = 'checkbox::v1' as const
  private static readonly dataSignature = {
    v1: { [ControlDataTypeKey]: this.v1DataType },
  } as const

  static readonly type = 'makeswift::controls::checkbox' as const

  static get schema() {
    const version = z.literal(1).optional()
    const versionedData = z.object({
      [ControlDataTypeKey]: z.literal(this.v1DataType),
      value: z.boolean(),
    })

    const schemas = <V, D>(value: SchemaType<V>, data: SchemaType<D>) => {
      const type = z.literal(this.type)

      const config = z.object({
        label: z.string().optional(),
        description: z.string().optional(),
        defaultValue: value,
      })

      const definition = z.object({
        type,
        config,
        version,
      })

      return {
        type,
        value,
        resolvedValue: value,
        data,
        versionedData,
        config,
        version,
        definition,
      }
    }

    return {
      version,
      relaxed: schemas(
        z.boolean().optional(),
        z.union([z.boolean(), versionedData, z.undefined()]),
      ),
      strict: schemas(z.boolean(), z.union([z.boolean(), versionedData])),
    }
  }

  static deserialize(data: DeserializedRecord): CheckboxDefinition {
    if (data.type !== Definition.type)
      throw new Error(
        `Checkbox: expected '${Definition.type}', got '${data.type}'`,
      )

    const { config, version } = Definition.schema.relaxed.definition.parse(data)
    return new CheckboxDefinition(config, version)
  }

  constructor(
    config: C,
    readonly version: z.infer<typeof Definition.schema.version>,
  ) {
    super(config)
  }

  get controlType() {
    return Definition.type
  }

  get schema(): ReturnedSchemaType<C> {
    return Definition.schema.relaxed
  }

  get dataSchema() {
    return (
      (this.config.defaultValue === undefined
        ? Definition.schema.relaxed
        : Definition.schema.strict) as Schema<C>
    ).data
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.dataSchema, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, ({ value }) => value)
      .otherwise((value) => value)
  }

  toData(value: ValueType<C>): DataType<C> {
    return match({ version: this.version, value })
      .with({ version: 1, value: P.boolean }, ({ value }) => ({
        ...Definition.dataSignature.v1,
        value,
      }))
      .with({ version: 1, value: undefined }, () => undefined)
      .with({ version: undefined }, ({ value }) => value)
      .otherwise(() => value)
  }

  copyData(
    data: DataType<C> | undefined,
    _context: CopyContext,
  ): DataType<C> | undefined {
    return data
  }

  resolveValue(
    data: DataType<C> | undefined,
  ): Resolvable<ResolvedValueType<C> | undefined> {
    return {
      name: Definition.type,
      readStable: () => this.fromData(data) ?? this.config.defaultValue,
      subscribe: () => () => {},
      triggerResolve: async () => {},
    }
  }

  createInstance(sendMessage: SendMessage<any>) {
    return new DefaultControlInstance(sendMessage)
  }

  accept<R>(visitor: ControlDefinitionVisitor<R>, ...args: unknown[]): R {
    return visitor.visitCheckbox(this, ...args)
  }
}

export class CheckboxDefinition<
  C extends Config = Config,
> extends Definition<C> {}

type UserConfig<D extends Config['defaultValue']> = Config & {
  defaultValue?: D
}

type NormedConfig<D extends Config['defaultValue']> = z.infer<
  SchemaByDefaultValue<D>['config']
>

export function Checkbox<D extends Config['defaultValue']>(
  config?: UserConfig<D>,
): CheckboxDefinition<NormedConfig<D>> {
  return new CheckboxDefinition((config ?? {}) as NormedConfig<D>, 1)
}
