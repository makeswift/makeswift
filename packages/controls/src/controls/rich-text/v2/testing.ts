import { z } from 'zod'

import { type DataType } from '../../associated-types'
import { type Resolvable } from '../../definition'
import {
  ControlInstance,
  DefaultControlInstance,
  type SendMessage,
} from '../../instance'

import { RichTextPluginControl } from './plugin'
import { RichTextDefinition } from './rich-text'

export const renderedNode = Symbol('rendered node')

export type RenderedNode = typeof renderedNode

type UserConfig = z.infer<typeof Definition.schema.userConfig>

class Definition extends RichTextDefinition<RenderedNode> {
  resolveValue(
    _data: DataType<RichTextDefinition<RenderedNode>> | undefined,
  ): Resolvable<RenderedNode | undefined> {
    return {
      name: Definition.type,
      readStable: () => renderedNode,
      subscribe: () => () => {},
      triggerResolve: async () => {},
    }
  }

  createInstance(sendMessage: SendMessage<any>): ControlInstance<any> {
    return new DefaultControlInstance(sendMessage)
  }

  toText(_data: DataType<RichTextDefinition<string>>): string {
    return ''
  }

  get pluginControls(): RichTextPluginControl[] {
    return []
  }

  pluginControlAt(_index: number) {
    return undefined
  }
}

export { Definition as RichTextDefinition }

export const RichText = (config?: UserConfig) => new Definition(config ?? {})

RichText.Mode = Definition.Mode
