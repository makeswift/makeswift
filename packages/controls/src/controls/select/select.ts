import { z } from 'zod'

import { map } from '../../lib/functional'
import { safeParse, unionOfLiterals, type ParseResult } from '../../lib/zod'

import { type CopyContext } from '../../context'
import {
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'

import {
  ControlDefinition,
  serialize,
  type Resolvable,
  type SchemaType,
} from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'
import { ControlDefinitionVisitor } from '../visitor'

type Option<T extends string> = { readonly value: T; readonly label: string }
type OptionList<T extends string> = readonly [Option<T>, ...Option<T>[]]

type Config<T extends string = string> = {
  readonly options: OptionList<T>
  readonly defaultValue?: T
  readonly label?: string
  readonly labelOrientation?: 'horizontal' | 'vertical'
  readonly description?: string
}

type ItemType<C extends Config> = C extends Config<infer Item> ? Item : never
type DataType<C extends Config> = undefined extends C['defaultValue']
  ? ItemType<C> | undefined
  : ItemType<C>

type ValueType<C extends Config> = DataType<C>
type ResolvedValueType<C extends Config> = ValueType<C>

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>
> {
  static readonly type = 'makeswift::controls::select' as const

  static schema<T, D, V>(
    item: SchemaType<T>,
    data: SchemaType<D>,
    defaultValue: SchemaType<V>,
  ) {
    const type = z.literal(Definition.type)

    const value = data
    const resolvedValue = value

    const config = z.object({
      options: z.array(z.object({ value: item, label: z.string() })).nonempty(),
      defaultValue,
      label: z.string().optional(),
      labelOrientation: z
        .union([z.literal('horizontal'), z.literal('vertical')])
        .optional(),
      description: z.string().optional(),
    })

    const definition = z.object({
      type,
      config,
    })

    return {
      type,
      data,
      value,
      resolvedValue,
      config,
      definition,
    }
  }

  static deserialize(data: DeserializedRecord): SelectDefinition {
    if (data.type !== Definition.type) {
      throw new Error(
        `Select: expected type ${Definition.type}, got ${data.type}`,
      )
    }

    const { config: configSchema } = Definition.schema(
      z.coerce.string(),
      z.coerce.string().optional(),
      z.coerce.string().optional(),
    )

    const def = z.object({
      type: z.literal(Definition.type),
      config: configSchema,
    })

    const { config } = def.parse(data)

    return new SelectDefinition(config as Config)
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    return Definition.schema(
      this.itemSchema,
      this.dataSchema.optional() as SchemaType<DataType<C>>,
      this.itemSchema.optional(),
    )
  }

  get itemSchema(): SchemaType<ItemType<C>> {
    return unionOfLiterals(
      map(this.config.options, ({ value }) => value as ItemType<C>),
    )
  }

  get dataSchema(): SchemaType<DataType<C>> {
    const item = this.itemSchema
    return (
      this.config.defaultValue === undefined ? item.optional() : item
    ) as SchemaType<DataType<C>>
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.dataSchema, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    return this.dataSchema.optional().parse(data)
  }

  toData(value: ValueType<C>): DataType<C> {
    return value
  }

  copyData(
    data: DataType<C> | undefined,
    _context: CopyContext,
  ): DataType<C> | undefined {
    return data
  }

  resolveValue(
    data: DataType<C> | undefined,
  ): Resolvable<ResolvedValueType<C> | undefined> {
    return {
      name: Definition.type,
      readStable: () =>
        this.fromData(data) ??
        (this.config.defaultValue as ResolvedValueType<C>),
      subscribe: () => () => {},
      triggerResolve: async () => {},
    }
  }

  createInstance(sendMessage: SendMessage) {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  accept<R>(visitor: ControlDefinitionVisitor<R>, ...args: unknown[]): R {
    return visitor.visitSelect(this, ...args)
  }
}

export class SelectDefinition<
  C extends Config = Config,
> extends Definition<C> {}

type UserConfig<
  T extends string,
  D extends Config<T>['defaultValue'],
> = Config<T> & {
  defaultValue?: D
}

type NormedConfig<
  T extends string,
  D extends Config<T>['defaultValue'],
> = Config<T> &
  (undefined extends D
    ? {
        readonly defaultValue?: T
      }
    : { readonly defaultValue: T })

export function Select<
  const T extends string,
  D extends Config<T>['defaultValue'],
>(
  config: UserConfig<T, D> & { readonly options: OptionList<T> },
): SelectDefinition<NormedConfig<T, D>> {
  return new SelectDefinition(config as NormedConfig<T, D>)
}
export { type Config as SelectConfig }
