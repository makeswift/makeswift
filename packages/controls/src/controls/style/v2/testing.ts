import {
  StyleV2Definition as BaseDefinition,
  StyleV2Config,
  StyleV2PropDefinition,
} from './style-v2'

export type StylesObject = {
  textAlignment?: 'left' | 'right' | 'center'
  visibility?: 'visible' | 'hidden'
}

type PropDefinition = StyleV2PropDefinition
type Config<Prop extends PropDefinition> = StyleV2Config<Prop, StylesObject>

export class Definition<
  Prop extends PropDefinition = PropDefinition,
> extends BaseDefinition<Prop, StylesObject> {}

export function StyleV2<Prop extends PropDefinition>(
  config: Config<Prop>,
): Definition<Prop> {
  return new Definition(config)
}

export { Definition as StyleV2Definition }
