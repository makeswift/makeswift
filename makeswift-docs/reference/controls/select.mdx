---
title: Select
description: Adds a Select option (dropdown) to a component in the Makeswift builder to visually select a `string` value.
icon: "list-dropdown"
---

<Frame
  type="glass"
  caption="A Select panel for a Button component to change the style"
>
  <img src="/images/select-panel.png" />
</Frame>

## Params

<ParamField query="label" type="string" default="Text">
  Text for the panel label in the Makeswift builder.
</ParamField>

<ParamField
  query="labelOrientation"
  type='"vertical" | "horizontal"'
  default="horizontal"
>
  Orientation of the dropdown label within the panel.
</ParamField>

<ParamField query="options" type="SelectOption<T>[]" required>
  An array of objects of type `SelectOption<T>` that contains the options available in the panel input.

```ts
type SelectOption<T extends string> = {
  value: T
  label: string
}
```

The `label` field is displayed in the Makeswift builder.

</ParamField>

<ParamField query="defaultValue" type="T">
  The value passed to your component when nothing is set in the Makeswift
  builder.
</ParamField>

## Prop type

The Select control passes the generic type `T` from the selected `Option` to your component. If you don't set a `defaultValue` and no value is set in the builder, your component will receive `undefined`.

## Example

The following example adds a Select control to the `variant` prop of a Button component.

<CodeGroup>

```tsx Button.makeswift.ts
import { Select } from "@makeswift/runtime/controls"

import { runtime } from "@/makeswift/runtime"

import { Button } from "./Button"

runtime.registerComponent(Button, {
  type: "button",
  label: "Button",
  props: {
    variant: Select({
      label: "Style",
      labelOrientation: "horizontal",
      options: [
        { value: "floating", label: "Floating" },
        { value: "outline", label: "Outline" },
        { value: "clear", label: "Clear" },
      ],
      defaultValue: "floating",
    }),
  },
})
```

```tsx Button.tsx
import clsx from "clsx"

interface Props {
  variant?: "floating" | "outline" | "clear"
}

export function Button({ variant = "floating" }) {
  return (
    <a
      className={clsx(
        "px-4 py-3 rounded",
        {
          floating: "bg-primary text-white shadow",
          outline: "bg-transparent border border-primary text-primary",
          clear: "text-primary bg-transparent border-none",
        }[variant]
      )}
      href={link.href}
      target={link.target}
    >
      Click me
    </a>
  )
}
```

</CodeGroup>

import NamingConventions from "/snippets/naming-conventions.mdx"

<NamingConventions />
