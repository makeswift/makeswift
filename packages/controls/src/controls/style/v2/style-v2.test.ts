import { Checkbox } from '../../checkbox'
import { Select } from '../../select'

import { StyleV2, StyleV2Definition } from './testing'

const textAlign = Select({
  label: 'Alignment',
  options: [
    {
      label: 'Left Align',
      value: 'left',
    },
    {
      label: 'Center Align',
      value: 'center',
    },
    {
      label: 'Right Align',
      value: 'right',
    },
  ],
  defaultValue: 'left',
})

const visibility = Checkbox({ defaultValue: true })

describe('StyleV2', () => {
  describe('constructor', () => {
    test('returns correct definition', () => {
      expect(
        StyleV2({
          type: textAlign,
          getStyle(textAlign) {
            return { textAlignment: textAlign }
          },
        }),
      ).toMatchSnapshot('text alignment')

      expect(
        StyleV2({
          type: visibility,
          getStyle(visibility) {
            return { visibility: visibility ? 'visible' : 'hidden' }
          },
        }),
      ).toMatchSnapshot('visibility')
    })

    test('disallows extraneous properties', () => {
      StyleV2({
        type: textAlign,
        getStyle(textAlign) {
          return { textAlignment: textAlign }
        },
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: StyleV2Definition) {}
    assignTest(
      StyleV2({
        type: textAlign,
        getStyle(textAlign) {
          return { textAlignment: textAlign }
        },
      }),
    )

    assignTest(
      StyleV2({
        type: visibility,
        getStyle(visibility) {
          return { visibility: visibility ? 'visible' : 'hidden' }
        },
      }),
    )
  })
})
