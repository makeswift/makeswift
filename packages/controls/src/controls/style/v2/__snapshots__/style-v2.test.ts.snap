// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleV2 constructor returns correct definition: text alignment 1`] = `
Definition {
  "config": {
    "getStyle": [Function],
    "type": SelectDefinition {
      "config": {
        "defaultValue": "left",
        "label": "Alignment",
        "options": [
          {
            "label": "Left Align",
            "value": "left",
          },
          {
            "label": "Center Align",
            "value": "center",
          },
          {
            "label": "Right Align",
            "value": "right",
          },
        ],
      },
    },
  },
}
`;

exports[`StyleV2 constructor returns correct definition: visibility 1`] = `
Definition {
  "config": {
    "getStyle": [Function],
    "type": CheckboxDefinition {
      "config": {
        "defaultValue": true,
      },
      "version": 1,
    },
  },
}
`;
